### ISSUE TITLE
Warning messages in Signac() and then Error in GenerateLabels()

### ISSUE THREAD

**Issue by levinhein** [mathewchamberlain/SignacX#10]
Hello. I'm trying out SignacX. 

**For Signac() function, I got 5 warning messages. Although, it did generate a "labels" object.**
Warning messages:
```
1: In if (class(E) == "matrix") { :
  the condition has length > 1 and only the first element will be used
2: In min(x) : no non-missing arguments to min; returning Inf
3: In max(x) : no non-missing arguments to max; returning -Inf
4: In min(x) : no non-missing arguments to min; returning Inf
5: In mclapply(X, function(...) { :
  all scheduled cores encountered errors in user code
```
Below is a picture of warning messages and how the generated labels look like.
(a) the 5 warning messages. You will also observe "0%, ETA NA"
![image](https://user-images.githubusercontent.com/68268222/123673831-7f173c80-d873-11eb-89c3-78e760c9c591.png)
(b) the table of "labels" object
![image](https://user-images.githubusercontent.com/68268222/123674135-d87f6b80-d873-11eb-8daf-a7d1fa4f0747.png)

I still tried to proceed with the next step, but now it got an error.

**For GenerateLabels() function, the error:**
`_Error in Cspdiagprod(x, y, boolArith = NA) : non-matching dimensions_`
![image](https://user-images.githubusercontent.com/68268222/123674388-25fbd880-d874-11eb-8653-ffe52aba4c08.png)

It seems like the issue is in dimensions? But I don't know how to resolve this. May I ask for your help please?

**Comment by mathewchamberlain** [comment]
Sure. A couple questions: first, when you run class(KD.combined), what does it say? Second, can you screenshot the results of KD.combined[1:3,1:3]? I just want to see what type of gene symbols you are using. SignacX expects row names to be labeled by gene symbols (HUGO).

**Comment by levinhein** [comment]
Wow, thank you so much for fast reply. Here's the screenshot of those two things:

![image](https://user-images.githubusercontent.com/68268222/123676654-ce12a100-d876-11eb-8641-0615d9c5a265.png)

**Comment by mathewchamberlain** [comment]
No problem. A few more questions: what is returned by names(KD.combined@graphs)? What version of Seurat / SignacX are you using? (you can get this info by running sessionInfo())

**Comment by mathewchamberlain** [comment]
Also, what is returned by DefaultAssay(KD.combined)?

**Comment by levinhein** [comment]
For names(KD.combined@graphs):

![image](https://user-images.githubusercontent.com/68268222/123677122-542ee780-d877-11eb-9485-cc36800aedd6.png)

Seurat is 4.0.3 and SignacX is 2.2.2


![image](https://user-images.githubusercontent.com/68268222/123677351-9526fc00-d877-11eb-9947-adee99ca273c.png)

For DefaultAssay(KD.combined), it's "integrated"

![image](https://user-images.githubusercontent.com/68268222/123677488-c1db1380-d877-11eb-9507-ff5120fda75e.png)

**Comment by mathewchamberlain** [comment]
Ah, I think that I see the problem. Could you try the SignacFast() function instead of Signac()?

The difference is only that SignacFast uses an ensemble of pre-trained neural networks, and it zero-imputes any genes that are not present in the target data. You have only 3,000 genes in your data; some of the signatures used by the Signac function have >1,000 genes. It might be getting down to 0 genes from some cell types (which are probably not present in your data), so it throws dimensionality errors.

Please let me know if this works, thanks. Signac() was not really designed for targeted panels of gene subsets; it's more designed for genome-wide scRNA-seq.

**Comment by levinhein** [comment]
**Here, SignacFast() does not work either.** 

![image](https://user-images.githubusercontent.com/68268222/123679957-b3dac200-d87a-11eb-8536-a89ba45ef130.png)



 **"You have only 3,000 genes in your data; some of the signatures used by the Signac function have >1,000 genes."**

Sorry for the naive question, what do you mean that I only have 3,000? Indeed, my 3, 000 is higher than 1, 000. 

**Targeted panels of gene subsets..**. I'm not sure whether the 8 datasets that I integrated are from targeted panels, but I don't think they are not genome-wide scale. Here are the original dimensions of the datasets that I integrated. As you may see, the number of genes are roughly 10k-14k. Also, during QC pre-processing the filter for nFeature_RNA (genes) is indeed just around 2k-5k depending on the dataset because too high number of genes are doublets (https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). I did filter my nFeature_RNA to ~3000 (varies on each of my 8 datasets). 

![image](https://user-images.githubusercontent.com/68268222/123680945-cdc8d480-d87b-11eb-8fa9-2498c549e7e7.png)

Also, during integration, the SelectIntegrationFeatures() in SCTransform method of scRNA-seq data integration, specifies exactly 3000 as the number of features. This could be the reason for my 3000 genes on my final integrated data.

`SelectIntegrationFeatures(object.list = KD.list, nfeatures = 3000)`

**Comment by mathewchamberlain** [comment]
Yes, I have a few more ideas. What I meant is that SignacX (2.2.2) detected only 3,000 genes in the data. By default, SignacX runs on this field of the Seurat object:
KD.list@assays[[default.assay]]@counts

Where default.assay is:
default.assay <- Seurat::DefaultAssay(KD.list)

I have never used the integrated Seurat objects; it seems that dim(KD.list) returns 3,000 genes and 9,401 cells. The 9,401 cells look right to me (based on the list you sent me), but the 3,000 genes seems to refer to the integration features across the data. I would probably run SignacX differently in this case. Can you try:

list_of_data = list(NKD1, M_A1A2, KD_A3, KD_A4, KD_A5, KD_A6, KD_R1, KD_R2); # this is a list of Seurat objects
lbls = lapply(list_of_data, function(x) {
d=SignacFast(x, num.cores = 4)
GenerateLabels(cr = d, E = x)
})

Then you will have run SignacX on each data set independently. You can then merge the data (using integration anchors), and add the metadata to the integrated object;

meta = do.call(rbind, lbls)
KD.list@metadata$celltypes = meta$CellTypes
